### PR TITLE
Merging GFS.v16 production branch changes to develop

### PR DESCRIPTION
**Pull request by ajhenrique**
* Adding OMP loop in wmgridmd for optimizing initialization time, speeds up by up to OMP_NUM_THREADS init time, 
* Removes the excessive esmf field gather in wmesmfmd.ftn, speeds up exchange of ATM->WAV data,
* Added OMP loop and allocatable variables to WMUPDV in call to interpolation subroutine: speeds up more expensive interpolation loop for curvilinear grids,
* Bug fix for WORD array within the context of defining second restart file stream.

### CODE REVIEW

**Comment by ajhenrique** [comment]
Here is a list of non-identical cases, the usual suspects.
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (10 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (9 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e_c                     (1 files differ)
mww3_test_04/./work_PR3_UQ_MPI_NC_b_c                     (2 files differ)
mww3_test_07/./work_PR3_UQ                     (3 files differ)
ww3_tp2.10/./work_MPI                     (7 files differ)
ww3_tp2.16/./work_MPI                     (6 files differ)
ww3_tp2.7/./work_ST0                     (1 files differ)
ww3_ts2/./work_ST4_PR3_UQ                     (1 files differ)

**Comment by ajhenrique** [comment]
Here is the output from matrix.comp
[matrix.zip](https://github.com/NOAA-EMC/WW3/files/4471573/matrix.zip)

**Review by aliabdolali — APPROVED** [review]
Looks good to me Henrique.

**Comment by ajhenrique** [comment]
I noticed I had not merged the latest develop into the GFS.v16 branch. After merging, I reran the non-identical cases and expectedly had:
**********************************************************************
************************ identical cases *****************************
**********************************************************************
ww3_tp2.10/./work_MPI
ww3_tp2.16/./work_MPI
ww3_ts2/./work_ST2_PR3_UQ

File: model/ftn/wmupdtmd.ftn [hunk 1]
```diff
@@ -1177,6 +1177,7 @@
                                  YGRDC(:,:), HPFACI(:,:), HQFACI(:,:) !RP
 
       INTEGER, POINTER        :: ICLOSE
+      real,allocatable :: XGRTMP(:),YGRTMP(:)
```
Comment (ukmo-ccbunney) [inline model/ftn/wmupdtmd.ftn]: Just a comment on code style - perhaps capitalize and space the "real,allocatable" statement to match code style in rest of module?

File: model/ftn/wmesmfmd.ftn [hunk 2]
```diff
@@ -4742,16 +4742,30 @@
...
-        do k = 0,npet-1
-          call ESMF_FieldGather( field,  floc1d, rootPet=k, vm=vm, rc=rc )
-          if (ESMF_LogFoundError(rc, PASSTHRU)) return
+        call ESMF_VMbroadcast( vm, bcstData=floc1dary, count=count, rootPet=0, rc=rc)
+        if (ESMF_LogFoundError(rc, PASSTHRU)) return
+        do j=1,n2
+          do i=1,n1
+            fout(i,j) = floc1dary(i+(j-1)*n1)
+          enddo
         enddo
```
Comment (ukmo-ccbunney) [inline model/ftn/wmesmfmd.ftn:4761]: Is `fout` not the same as `floc` here (as calculated on line 4752)? I.e. can these 5 lines not simply be replaced by:
```
    fout = floc
```

Or does  `ESMF_VMbroadcast` potentially change the `floc1dary` array (I don't know where this library is)?

**Review by ukmo-ccbunney — COMMENTED** [review]
@ajhenrique Looks good - just a few comments below.

File: model/ftn/wmesmfmd.ftn (hunk 2 above)
Comment (ajhenrique) [reply model/ftn/wmesmfmd.ftn:4761]: @junwang-noaa this is a question for you. Please comment.

File: model/ftn/wmupdtmd.ftn (hunk 1 above)
Comment (ajhenrique) [reply model/ftn/wmupdtmd.ftn]: @ukmo-ccbunney this slipped from my review of changes, good eye! The change has been made.

File: model/ftn/wmesmfmd.ftn (hunk 2 above)
Comment (junwang-noaa) [reply model/ftn/wmesmfmd.ftn:4761]: The ESMF_VMbroadcast is in ESMF lib. Also we don't recommend to match different rank of array using pointer as there might be machine dependent on how the 1D array is stored, it's not safe. Also we want to add OPENMP on that loop later

Comment (ukmo-ccbunney) [reply model/ftn/wmesmfmd.ftn:4761]: @junwang-noaa: `fout` and `floc` are the same rank as far as I can see. Both rank 2 and shape (n1, n2). I don't have a strong opinion about the comment I made, I just thought that it looked like an unnecessary step and you might be able to save some CPU cycles (admittedly it's probably a trivial amount of cycles!)

**Comment by junwang-noaa** [comment]
Sorry I need to clarify. I thought we are talking about floc1dary and fout,
floc1dary is a 1D array (rank 1), floc (rank 2, shape (21,n2)) is only
available on root PE, while floc1dary is available on all the mpi tasks. I
don't think we can do fout=floc, the code will break.

File: model/ftn/wmesmfmd.ftn (hunk 2 above)
Comment (ajhenrique) [reply model/ftn/wmesmfmd.ftn:4761]: This discussion may be continued in the comments below.

Comment (ukmo-ccbunney) [reply model/ftn/wmesmfmd.ftn:4761]: >Sorry I need to clarify. I thought we are talking about floc1dary and fout,
floc1dary is a 1D array (rank 1), floc (rank 2, shape (21,n2)) is only
available on root PE, while floc1dary is available on all the mpi tasks. I
don't think we can do fout=floc, the code will break.

@jianwang-noaa Sorry - I'm obviously not understanding some subtlety here. I'll approve this review to expedite it's implementation and we can discuss this a bit more offline. Cheers!

**Review by ukmo-ccbunney — APPROVED** [review]
Happy with this. All good from me!

**Comment by ajhenrique** [comment]
Thanks @aliabdolali @ukmo-ccbunney @junwang-noaa for the comments/discussions and approvals! This PR will be merged to develop now.

**Pull request merged**